### DISCUSSION TITLE
Wi-Fi signal quality parameters - confused interpretation

### DISCUSSION THREAD

**Discussion by zokl** [arendst/Tasmota#17410] [Ideas]
### PROBLEM DESCRIPTION

Hello,

While analyzing data obtained via MQTT, I encountered a problem interpreting Wi-Fi parameters RSSI and Signal. From the values, it seems that both parameters are swapped. RSSI is Receive Signal Strength Indicator and should be reported in dBm, and it is always a negative number. As for the second item, it seems to take values from 0 to 100, and it is probably the CCQ - Client Connection Quality. This parameter is commonly called only connection quality or simple quality. 

Is that correct? Is it possible to correct and relabel it?

References: 
CCQ - https://wiki.mikrotik.com/wiki/Manual:Wireless_FAQ#What_is_CCQ_and_how_are_the_values_determined? 
RSSI - https://en.wikipedia.org/wiki/Received_signal_strength_indication

### REQUESTED INFORMATION
_Make sure your have performed every step and checked the applicable boxes before submitting your issue. Thank you!_

- [x] Read the [Contributing Guide and Policy](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md) and [the Code of Conduct](https://github.com/arendst/Tasmota/blob/development/CODE_OF_CONDUCT.md)
- [x] Searched the problem in [issues](https://github.com/arendst/Tasmota/issues)
- [x] Searched the problem in [discussions](https://github.com/arendst/Tasmota/discussions)
- [x] Searched the problem in the [docs](https://tasmota.github.io/docs/FAQ)
- [x] Searched the problem in the [chat](https://discord.gg/Ks2Kzd4)
- [x] Device used (e.g., Sonoff Basic): all ESP8266 ESP32
- [ ] Tasmota binary firmware version number used: 12.2.0.6
  - [ ] Pre-compiled
  - [x] Self-compiled
- [ ] Flashing tools used: _____
- [ ] Provide the output of command: `Backlog Template; Module; GPIO 255`:
```lua
  Configuration output here:


```
- [ ] If using rules, provide the output of this command: `Backlog Rule1; Rule2; Rule3`:
```lua
  Rules output here:


```
- [ ] Provide the output of this command: `Status 0`:
```lua
  STATUS 0 output here:


```
- [ ] Set `weblog` to 4 and then, when you experience your issue, provide the output of the Console log:
```lua
  Console output here:

esp32:

14:58:01.910 MQT: tele/TemesIoT_33D858/STATE = {"Time":"2022-12-16T14:58:01","Uptime":"0T04:40:15","UptimeSec":16815,"Heap":146,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":43},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"TemesIoT","BSSId":"2C:C8:1B:63:D6:19","Channel":8,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:10"}}
14:58:01.920 MQT: tele/TemesIoT_33D858/SENSOR = {"Time":"2022-12-16T14:58:01","Switch1":"ON"}
15:03:01.883 MQT: tele/TemesIoT_33D858/STATE = {"Time":"2022-12-16T15:03:01","Uptime":"0T04:45:15","UptimeSec":17115,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":43},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"TemesIoT","BSSId":"2C:C8:1B:63:D6:19","Channel":8,"Mode":"11n","RSSI":92,"Signal":-54,"LinkCount":1,"Downtime":"0T00:00:10"}}
15:03:01.892 MQT: tele/TemesIoT_33D858/SENSOR = {"Time":"2022-12-16T15:03:01","Switch1":"ON"}
15:08:01.870 MQT: tele/TemesIoT_33D858/STATE = {"Time":"2022-12-16T15:08:01","Uptime":"0T04:50:15","UptimeSec":17415,"Heap":149,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"Berry":{"HeapUsed":3,"Objects":43},"POWER1":"OFF","POWER2":"OFF","POWER3":"OFF","POWER4":"OFF","Wifi":{"AP":1,"SSId":"TemesIoT","BSSId":"2C:C8:1B:63:D6:19","Channel":8,"Mode":"11n","RSSI":88,"Signal":-56,"LinkCount":1,"Downtime":"0T00:00:10"}}
15:08:01.880 MQT: tele/TemesIoT_33D858/SENSOR = {"Time":"2022-12-16T15:08:01","Switch1":"ON"}

esp8266:
02:49.385 MQT: tele/Kotelna_CS300/STATE = {"Time":"2022-12-16T15:02:49","Uptime":"0T20:00:20","UptimeSec":72020,"Heap":18,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Dum300-1","BSSId":"B8:69:F4:E7:A5:DC","Channel":1,"Mode":"11n","RSSI":36,"Signal":-82,"LinkCount":1,"Downtime":"0T00:00:06"}}
15:02:49.470 MQT: tele/Kotelna_CS300/SENSOR = {"Time":"2022-12-16T15:02:49","ANALOG":{"MQ2_0":3.36},"DS18B20-1":{"Id":"00000E68984D","Temperature":47.3},"DS18B20-2":{"Id":"00000E9495AD","Temperature":52.6},"DS18B20-3":{"Id":"00000E95AA10","Temperature":47.5},"DS18B20-4":{"Id":"0004334EAFFF","Temperature":49.1},"AM2301":{"Temperature":24.9,"Humidity":19.7,"DewPoint":0.1},"TempUnit":"C"}
15:07:49.485 MQT: tele/Kotelna_CS300/STATE = {"Time":"2022-12-16T15:07:49","Uptime":"0T20:05:20","UptimeSec":72320,"Heap":18,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":1,"POWER1":"OFF","POWER2":"OFF","Wifi":{"AP":1,"SSId":"Dum300-1","BSSId":"B8:69:F4:E7:A5:DC","Channel":1,"Mode":"11n","RSSI":32,"Signal":-84,"LinkCount":1,"Downtime":"0T00:00:06"}}
15:07:49.638 MQT: tele/Kotelna_CS300/SENSOR = {"Time":"2022-12-16T15:07:49","ANALOG":{"MQ2_0":3.31},"DS18B20-1":{"Id":"00000E68984D","Temperature":47.3},"DS18B20-2":{"Id":"00000E9495AD","Temperature":52.1},"DS18B20-3":{"Id":"00000E95AA10","Temperature":47.4},"DS18B20-4":{"Id":"0004334EAFFF","Temperature":48.8},"AM2301":{"Temperature":24.9,"Humidity":20.0,"DewPoint":0.3},"TempUnit":"C"}

```

### TO REPRODUCE
_Steps to reproduce the behavior:_

### EXPECTED BEHAVIOUR
_A clear and concise description of what you expected to happen._

### SCREENSHOTS
_If applicable, add screenshots to help explain your problem._

### ADDITIONAL CONTEXT

**(Please, remember to close the issue when the problem has been addressed)**

**Comment by sfromis** [comment]
They are the way they are for historical reasons, it would be a breaking change to update. Originally, RSSI was transformed to a "human-friendly" value, with the "Signal" added later for people preferring the dBm value directly. Obviously, we are well aware of what the definitions are, but this is not going to be changed.

**Reply by zokl** [reply]
> They are the way they are for historical reasons, it would be a breaking change to update. Originally, RSSI was transformed to a "human-friendly" value, with the "Signal" added later for people preferring the dBm value directly. Obviously, we are well aware of what the definitions are, but this is not going to be changed.

The historical reason is a very bad argument for the future extension. I think it will be good to use RSSI as a signal indicator (dBm) and a Signal as a „human-friendly” value. The current solution is the very confusing solution for integration :-(.

**Reply by Jason2866** [reply]
And it is not nice to brake existing integrations with a change. The dBm value is not a verified value. It is what espressif is telling us via API. Real value? We dont know. So all just indicators....
In real life espressif devices are making trouble not because rssi is not good, the problem is the closed wifi stack which has bugs.... Ask ASUS Router users ;-)
btw. your links are not pointing to an ISO/IEC  So our way is not proved wrong.
Even your posted link says
```
There is no standardized relationship of any particular physical parameter to the RSSI reading. The 802.11 standard does not define any relationship between RSSI value and power level in [milliwatts](https://en.m.wikipedia.org/wiki/Milliwatts) or [decibels referenced to one milliwatt (dBm)](https://en.m.wikipedia.org/wiki/DBm). Vendors and chipset makers provide their own accuracy, granularity, and range for the actual power (measured as milliwatts or decibels) and their range of RSSI values (from 0 to RSSI maximum).[[2]](https://en.m.wikipedia.org/wiki/Received_signal_strength_indication#cite_note-2) One subtlety of the 802.11 RSSI metric comes from how it is sampled—RSSI is acquired during only the preamble stage of receiving an 802.11 frame, not over the full frame.[[3]](https://en.m.wikipedia.org/wiki/Received_signal_strength_indication#cite_note-:0-3)
```
It can be everything.

**Reply by sfromis** [reply]
Within computing, I see tons of small anomalies, and have gotten used to being watchful instead of being confused. Things which I with the convenience of the rear view mirror would have wanted to have been done differently, but not things to bitch about.

**Reply by zokl** [reply]
Most radio technologies I know maintain RSSI in dBm as a power variable. For Wi-Fi I know this is also the case. Yes, it's not standard, but it's common practice.

**Reply by szczukot** [reply]
In my opinion, the explanation that it would be a breaking change is weak. You can't go wrong with "mistake" because "it's always been that way"
Today, I also thinkung what 
"RSSI":64,"Signal":-68
 is meaning.
 
 Especially since I have something like this:
 https://github.com/arendst/Tasmota/discussions/21370

**Reply by Jason2866** [reply]
Open Source change like you want.

**Reply by sfromis** [reply]
It means that it is sort of 64%, with the dBm value of -68. And backwards compatibility is a major concern when a change could wreck havoc on existing users. This is very different from "it's always been that way".

Thus, this is is a **strong** explanation, even if you may not like it.

The specific calculation is quality = 2 * (dBm + 100), capped at values 0..100. Hence, dBm values of -50 or better gets to be 100, 0 is only when extremely horrible at -100 or worse, while a quality about 50 is about the minimum usable signal strength, -75 dBm.

I suspect that the quality conversion of the dBm value was due to Domoticz seemingly expecting "RSSI" to be a signal quality from 0..10, or around that.

**Reply by szczukot** [reply]
In every software has "breaking changes". 
The main/major version is then bumped, these changes will be recorded in the changelog, and it is up to the user what to do with them.
This is what development and bug fixing is all about.

**Reply by s-hadinger** [reply]
To be clear, we will not introduce a breaking change in the API (json) without a bigger benefit.

We know that the naming is not ideal, but we have to live with it. The only option is to create a new attribute with a new (correct) name. But we have no short term plan to add 2 more parameters.
